ll2_global_planner/src/ll2_global_planner_nodelet.cpp

```diff
@@ -44,10 +49,178 @@ void Ll2GlobalPlannerNl::loadParams()
+
+  if (continuous_lane.size() != shortest_path.size())
+  {
+    ROS_WARN("This route contains a lane change which is currently unsupported");
+    return;
+  }
+
+  ROS_INFO("Found a path containing %lu lanelets", shortest_path.size());
+
+  // Convert to autoware waypoints
+  std::vector<autoware_msgs::Waypoint> waypoints;
+
+  // Loop over each lanelet
+  for (auto& lanelet : continuous_lane.lanelets())
+  {
+    const std::string turn_direction = lanelet.attributeOr("turn_direction", "straight");
+    traffic_rules::SpeedLimitInformation speed_limit = traffic_rules->speedLimit(lanelet);
+    ConstLineString3d centerline = lanelet.centerline();
+    int wp_length = centerline.size() - 1;
+
+    // Loop over each centerline point
+    for (int i = 0; i <= wp_length; i++)
+    {
+      auto point = centerline[i];
+
+      autoware_msgs::Waypoint new_wp;
+      new_wp.pose.pose.position.x = point.x();
+      new_wp.pose.pose.position.y = point.y();
+      new_wp.pose.pose.position.z = point.z();
+      new_wp.twist.twist.linear.x = speed_limit.speedLimit.value(); // m/s
+
+      int steering_state = autoware_msgs::WaypointState::STR_STRAIGHT;
+      if (turn_direction.compare("right") == 0)
+      {
+        steering_state = autoware_msgs::WaypointState::STR_RIGHT;
+      }
+      if (turn_direction.compare("left") == 0)
+      {
+        steering_state = autoware_msgs::WaypointState::STR_LEFT;
+      }
+      new_wp.wpstate.steering_state = steering_state;
+
+      waypoints.push_back(new_wp);
+    }
+  }
+
+  int wp_id = 0;
+  bool first_point = true;
+  float smallest_goal_dist = std::numeric_limits<float>::infinity();
+  int smallest_goal_wp_id = 0;
+  autoware_msgs::Waypoint* prev_waypoint = nullptr;
+  std::vector<autoware_msgs::Waypoint> processed_waypoints;
+
+  // Remove duplicates and set waypoint orientations
+  for (autoware_msgs::Waypoint& waypoint : waypoints)
+  {
+    geometry_msgs::Pose& pose = waypoint.pose.pose;
+    geometry_msgs::Pose& prev_pose = prev_waypoint->pose.pose;
+
+    if (!first_point)
+    {
+      // Check for duplicate points
+      float dist = std::hypot(pose.position.x - prev_pose.position.x, pose.position.y - prev_pose.position.y);
+      if (dist < duplicate_dist_threshold)
+      {
+        continue;
+      }
+
+      // Set orientation of the previous point, aiming at current point
+      float yaw = std::atan2(pose.position.y - prev_pose.position.y, pose.position.x - prev_pose.position.x);
+
+      tf2::Quaternion orientation;
+      orientation.setRPY(0, 0, yaw);
+      tf2::convert(orientation, prev_pose.orientation);
+    }
+    else
+    {
+      first_point = false;
+    }
+
+    // Calculate distance to goal point
+    float goal_dist = std::hypot(std::hypot(goal_point.x - pose.position.x, goal_point.y - pose.position.y), goal_point.z - pose.position.z);
+    if (goal_dist < smallest_goal_dist)
+    {
+      smallest_goal_dist = goal_dist;
+      smallest_goal_wp_id = wp_id;
```

Comment:
As above waypoint closest to goal may still be far away from goal location. Not sure how this is dealt with later.

Reply:
Since centerline points are spaced 1 meter apart, the closest waypoint is effectively the goal point projected onto the path.

Reply:
Well there maybe still be problems with a specific local planner - but I guess not so important if documented.